### ISSUE TITLE
sup-mail crashes on startup with ancient config. 

### ISSUE THREAD

**Issue by markstos** [sup-heliotrope/sup#516]
I installed `sup-mail` on  Ubuntu 16.04, ran `sup-mail` for the first time, and it crashed on startup. Below is `.sup/exception-log.txt`.

I believe the issue is apparently I tried `sup` in 2009, and still had a `.sup/config.yml` from then. It would be nice if `sup-mail` could not crash in the presence of a bad config file. 

```
--- NoMethodError from thread: main
undefined method `empty?' for Redwood::SentLoader:Class
/usr/lib/ruby/2.3.0/psych/class_loader.rb:26:in `load'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:396:in `resolve_class'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:160:in `visit_Psych_Nodes_Mapping'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:330:in `block in register_empty'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:330:in `each'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:330:in `register_empty'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:141:in `visit_Psych_Nodes_Sequence'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:311:in `visit_Psych_Nodes_Document'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:16:in `visit'
/usr/lib/ruby/2.3.0/psych/visitors/visitor.rb:6:in `accept'
/usr/lib/ruby/2.3.0/psych/visitors/to_ruby.rb:32:in `accept'
/usr/lib/ruby/2.3.0/psych/nodes/node.rb:38:in `to_ruby'
/usr/lib/ruby/2.3.0/psych.rb:253:in `load'
/usr/lib/ruby/2.3.0/psych.rb:471:in `block in load_file'
/usr/lib/ruby/2.3.0/psych.rb:471:in `open'
/usr/lib/ruby/2.3.0/psych.rb:471:in `load_file'
/usr/lib/ruby/vendor_ruby/sup.rb:145:in `load_yaml_obj'
/usr/lib/ruby/vendor_ruby/sup/source.rb:227:in `load_sources'
/usr/lib/ruby/vendor_ruby/sup/util.rb:610:in `method_missing'
/usr/lib/ruby/vendor_ruby/sup/index.rb:102:in `load'
/usr/lib/ruby/vendor_ruby/sup/util.rb:610:in `method_missing'
/usr/bin/sup-mail:144:in `<module:Redwood>'
/usr/bin/sup-mail:74:in `<main>'
```

Here's the old .sup/config.yaml which triggered the exception:

```
--- 
:confirm_no_attachments: true 
:editor: /usr/bin/vim -f -c 'setlocal spell spelllang=en_us' -c 'set filetype=mail'
:thread_by_subject: false
:confirm_top_posting: true
:accounts: 
  :default: 
    :sendmail: /usr/sbin/sendmail -oem -ti
    :signature: /home/mark/.signature
    :email: mark@stosberg.com
    :name: Mark Stosberg
    :alternates: []

:edit_signature: false
:ask_for_cc: true
:ask_for_bcc: false
:ask_for_subject: true
```

**Comment by egils-verkstad** [comment]
I also had sup 0.22.1 crash today because of invalid paths configured in the config-file. I would suggest to check the paths on startup and exit clean with a sensible errormessage if they are not valid.

**Comment by danc86** [comment]
The stack trace is from loading `sources.yaml` but the issue description only includes `config.yaml` so it's hard to be sure exactly what was going wrong.

This is likely related to the Syck-to-Psych transition which Ruby went through in 1.9 -> 2.0. From version 0.14.0 onwards, sup included a tool `sup-psych-ify-config-files` to handle the migration, but from version 0.16.0 onwards the tool refuses to run on Ruby 2.1+, presumably because Psych would no longer successfully parse the old Syck-generated YAML at that point.

At this point anybody hitting this issue would presumably be best to just re-create their `sources.yaml` from scratch.

**Comment by ulidtko** [comment]
... Still happening, BTW.

Edit `sources.yaml` (not `config.yaml`) — and comment out the `- !supmua.org,2006-10-01/Redwood/SentLoader {}` line.

**Comment by danc86** [comment]
Thanks for this extra info. I can see now how this would have gone wrong during the old migration.

There are a combination of a few different issues reported here but I agree sup could handle all of these more gracefully, so I've reopened this bug to look at making sup a bit more robust in the face of these old configs.

**Comment by danc86** [comment]
I did a bit more digging into the history around `sources.yaml` handling and `SentLoader` and I am still a bit uncertain...

I cannot find any version of the code which would have written out `sources.yaml` containing an instance of `SentLoader`. I went back to 0.22.1 and even in this version, `sup-config` will not write out `SentLoader` into `sources.yaml`. `sup-add` also does not accept `sup://sent` as a source to be added either. But there have been plenty of bug reports about old configurations where `SentLoader` exists in `sources.yaml` so obviously some very old sup versions were producing such configurations. My best guess is that it is related to 797af34eb246244c88b1923ef907c26f6de6ea43 and maybe before the point, `SentLoader` would be implicitly written back out to `sources.yaml`. But that point in the history is so far in the past I can no longer even run the code to confirm.

I also could not find any reproducer for the original ``undefined method `empty?' for Redwood::SentLoader:Class`` exception reported in this ticket. But I think this is just the same old issue with malformed tag URIs I fixed a long time back in 1aac9d43889f0f4f4420d3632f6cc8e7ac8cd639.

The issue with `- !supmua.org,2006-10-01/Redwood/SentLoader {}` is straightforward though. Commit 1aac9d43889f0f4f4420d3632f6cc8e7ac8cd639 just used a hacky regex to fix these invalid tag URIs but the regex did not account for the possibility of an empty map after the tag. I've fixed that in commit 7580318c9d58be5a24b320c0d69a87a319292d2d.

**Comment by ulidtko** [comment]
Hey @danc86, thanks for taking a look!

Frankly, I wasn't expecting such a thorough analysis in response (all the more appreciated) 😅 The workaround for me was simple enough — just comment out a line in that yaml. (And turn it into an "oh, where did my sent emails go?" problem — but later.)

Two things I'd like to say.

One: aggregators like `pkgs.org`, `repology.org` can shed some light onto the question "where do those ancient versions come from" — e.g. check this out: https://repology.org/project/sup-mail/versions 

For a simple example, the still currently latest LTS of Ubuntu, 24.04 Noble Numbat — includes `sup` at version **1.1** in its repositories. Meaning, when users do `apt install sup-mail`, that's the version they get. Which is by far not the latest, as we both know.

And of course, it's super common for servers to lag behind, even on LTS lines. Ubuntu 18 is still very much around. (Same would apply to every other distro.) Then, one easy-to-imagine scenario would go like this:
* user apt-installs sup on an old OS, tries it out, configures;
* sooner or later, hits issues, realizes it's an old version;
* upgrades sup, by reinstalling from rubygems or whatnot;
* *forgets the config in ~/.sup*;
* runs latest sup with "ancient" config (which may really be minutes old).

So yeah... it's a rather annoying problem.

---------

Two: I tried to resume my experiments with `sup` recently, but had to postpone them once again. Besides the issue with `SentLoader` — next there was *another* crash on start.

At that point I forced my way through it, with something like `cd /usr/lib/ruby/vendor_ruby/sup/ && curl https://github.com/sup-heliotrope/sup/commit/d2a2a0d7544237a29bf932773cc85e9050bd9c26.diff | patch -p3` but you know... that's not great. For all sorts of reasons. Leaves a bitter taste.

Not sure if telling all this is any useful to you — but at best, I hope this comment provides a fresh viewpoint from a power-user (or rather, would-be user). Not asking for any particular change; with all due respect, this is FLOSS, and I don't have any specific patch to propose. Yet please, give it some thought.

**Comment by danc86** [comment]
For sure, I understand that some distros will still ship old versions and a user may have a config left on disk from one of these old versions of sup. Version 1.1 is not that old though, I released that only a few years back. :-) This original bug report was filed in 2016 and the config containing `SentLoader` must have come from some version much older than that, but as I wrote above I cannot easily run any versions earlier than 0.22 anymore (released in 2015). I just wanted to have a reproducer to understand the bug properly.

To your other point. I'm not sure what version of Ruby shipped in Ubuntu 24.04. As you found, I did a few fixes for compatibility with Ruby ~2.2~ 3.2 and Ruby ~2.3~ 3.3 in recent years, it would be a shame if Ubuntu 24.04 had shipped with one of these newer Rubies but Sup was missing the fixes. But of course bugs happen, it is not the end of the world. Sup is intended to be easy to `gem install` or run from a git checkout and if you are using a distro that has not kept their Sup package well maintained, I would recommend this instead.

Regardless it is getting quite off-topic for this old bug. :-) Please file a fresh bug if you run into any crashes not covered by an existing bug report.
